src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp

```diff
@@ -58,56 +58,52 @@ G1ConcurrentRefineThread* G1ConcurrentRefineThreadControl::create_refinement_thr
   return result;
 }
 
-G1ConcurrentRefineThreadControl::G1ConcurrentRefineThreadControl() :
+G1ConcurrentRefineThreadControl::G1ConcurrentRefineThreadControl(uint max_num_threads) :
   _cr(nullptr),
-  _threads(nullptr),
-  _max_num_threads(0)
+  _threads(max_num_threads)
 {}
 
 G1ConcurrentRefineThreadControl::~G1ConcurrentRefineThreadControl() {
-  if (_threads != nullptr) {
-    for (uint i = 0; i < _max_num_threads; i++) {
-      G1ConcurrentRefineThread* t = _threads[i];
-      if (t == nullptr) {
-#ifdef ASSERT
-        for (uint j = i + 1; j < _max_num_threads; ++j) {
-          assert(_threads[j] == nullptr, "invariant");
-        }
-#endif // ASSERT
-        break;
-      } else {
-        delete t;
-      }
+  while (_threads.is_nonempty()) {
+    delete _threads.pop();
+  }
+}
+
+bool G1ConcurrentRefineThreadControl::ensure_threads_created(uint worker_id, bool initializing) {
+  assert(worker_id < max_num_threads(), "precondition");
+
+  while ((uint)_threads.length() <= worker_id) {
+    G1ConcurrentRefineThread* rt = create_refinement_thread(_threads.length(), initializing);
+    _threads.push(rt);
+
+    if (rt == nullptr) {
+      _threads.pop();
+      return false;
     }
-    FREE_C_HEAP_ARRAY(G1ConcurrentRefineThread*, _threads);
   }
+
+  return true;
 }
 
-jint G1ConcurrentRefineThreadControl::initialize(G1ConcurrentRefine* cr, uint max_num_threads) {
+jint G1ConcurrentRefineThreadControl::initialize(G1ConcurrentRefine* cr) {
   assert(cr != nullptr, "G1ConcurrentRefine must not be null");
   _cr = cr;
-  _max_num_threads = max_num_threads;
 
-  if (max_num_threads > 0) {
-    _threads = NEW_C_HEAP_ARRAY(G1ConcurrentRefineThread*, max_num_threads, mtGC);
-
-    _threads[0] = create_refinement_thread(0, true);
-    if (_threads[0] == nullptr) {
+  if (max_num_threads() > 0) {
+    _threads.push(create_refinement_thread(0, true));
+    if (_threads.at(0) == nullptr) {
       vm_shutdown_during_initialization("Could not allocate primary refinement thread");
       return JNI_ENOMEM;
     }
 
     if (UseDynamicNumberOfGCThreads) {
-      for (uint i = 1; i < max_num_threads; ++i) {
-        _threads[i] = nullptr;
+      for (uint i = 1; i < max_num_threads(); ++i) {
+        _threads.push(nullptr);
```

Comment:
Now that size/length is being used to track how many threads are in _threads, this is incorrect.  This fills
_threads with nulls, with length == max_num_threads at the end.  So later calls to ensure_threads_created
will do nothing.  I'm surprised this doesn't lead to crashes when attempting to activate threads other than
the primary thread.